src/extdata.rs

```diff
@@ -1,167 +1,190 @@
-use libdoodle::bpk1::{BPK1File, letter::Letter};
 
-macro_rules! handle_error {
-    ($res: expr) => {
-        let res = $res;
-        if R_FAILED(res) {
-            panic!("Error {res}");
-        }
-    };
-}
+use crate::error::panic_if_failed;
 
-pub fn read() -> impl Iterator<Item = (FS_DirectoryEntry, String, Letter)> {
-    // returns (file_path, vec<u8>)
-    let extdata_handle: FS_Archive = open_title_extdata(MEDIATYPE_SD, 0x00040000001A2E00).unwrap();
-
-    println!("handle is {}", extdata_handle);
-    list_dir(extdata_handle, "/letter/0000\0")
-        .into_iter()
-        .map(move |v| {
-            let filename = String::from_utf16(&v.name).unwrap();
-            let filename = format!("/letter/0000/{filename}\0");
-            let file = read_file(extdata_handle, &filename);
-            let letter = Letter::new_from_bpk1_bytes(&file).unwrap();
-            (v, filename, letter)
-        })
+pub enum SwapdoodleRegion {
+    EU,
+    US,
+    JP,
 }
 
-struct DirectoryIterator {
-    handle: Handle,
+pub struct ExtdataArchive {
+    pub region: SwapdoodleRegion,
+    pub archive: FS_Archive,
 }
 
-impl Iterator for DirectoryIterator {
-    type Item = FS_DirectoryEntry;
-
-    fn next(&mut self) -> Option<Self::Item> {
-        let mut read: u32 = 0;
-        let entry = unsafe {
-            let mut entry: FS_DirectoryEntry = mem::zeroed();
-            handle_error!(FSDIR_Read(
-                self.handle,
-                &mut read as *mut _,
-                1,
-                &mut entry as *mut _,
-            ));
-            entry
+impl ExtdataArchive {
+    pub fn open(region: SwapdoodleRegion) -> Result<ExtdataArchive, ()> {
+        let title_id: u64 = match region {
+            SwapdoodleRegion::EU => 0x00040000001A2E00,
+            SwapdoodleRegion::US => 0x00040000001A2D00,
+            SwapdoodleRegion::JP => 0x00040000001A2C00,
         };
-        if read == 1 { Some(entry) } else { None }
+        let extdata = (title_id as u32) >> 8;
+        let path: [u32; 3] = [MEDIATYPE_SD.into(), extdata, 0];
+
+        unsafe {
+            let mut extdata_handle: FS_Archive = 0;
+
+            match R_SUCCEEDED(FSUSER_OpenArchive(
+                &mut extdata_handle as *mut _,
+                0x00000006, // ARCHIVE_EXTDATA
+                FS_Path {
+                    type_: PATH_BINARY,
+                    size: 12,
+                    data: &path as *const _ as *const c_void,
+                },
+            )) {
+                true => Ok(ExtdataArchive {
+                    region,
+                    archive: extdata_handle,
+                }),
+                false => Err(()),
+            }
+        }
     }
-}
 
-impl Drop for DirectoryIterator {
-    fn drop(&mut self) {
+    pub fn read_file(&self, path: &str) -> Vec<u8> {
+        const BATCH_SIZE: u32 = 1024;
+
+        let mut file = Vec::<u8>::new();
+
+        let mut path: Vec<u16> = path.encode_utf16().collect();
+        path.push(0);
```

Comment:
```suggestion
        let path: Vec<u16> = path.encode_utf16().chain([0]).collect();
```
This seems like it would cause fewer allocations
Since it can collect the whole thing, including the null byte, in one go, instead of possibly having to extend the capacity of the vec during that push
([Compiler Explorer seems to back me up on this](https://godbolt.org/z/o998f5cbj))